diff --git a/setup.py b/setup.py
index cbe7a01..ff7e91f 100644
--- a/setup.py
+++ b/setup.py
@@ -72,8 +72,7 @@ try:
         package_dir={'ontquery':'export'},
         packages=['ontquery', 'ontquery.plugins'],
         python_requires='>=3.6',
-        setup_requires=['pytest-runner'],
-        tests_require=['pytest', 'rdflib', 'requests'],
+        tests_require=['pytest', 'pytest-runner', 'rdflib', 'requests'],
         install_requires=[
         ],
         extras_require={'dev':['pyontutils',],
