From 10ddee4cf03c370f085511066b8a37fa9fc803c8 Mon Sep 17 00:00:00 2001
From: Tom Gillespie <tgbugs@gmail.com>
Date: Mon, 16 Jan 2023 16:47:42 -0500
Subject: [PATCH] cache.cache remove imports depending on protobuf

---
 pennsieve/cache/cache.py | 4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/pennsieve/cache/cache.py b/pennsieve/cache/cache.py
index 635b720..d039559 100644
--- a/pennsieve/cache/cache.py
+++ b/pennsieve/cache/cache.py
@@ -18,7 +18,7 @@ from pennsieve.extensions import require_extension
 from pennsieve.models import DataPackage, TimeSeriesChannel
 from pennsieve.utils import usecs_since_epoch, usecs_to_datetime
 
-from .cache_segment_pb2 import CacheSegment
+# from .cache_segment_pb2 import CacheSegment  # no protobuf
 
 logger = log.get_logger("pennsieve.cache")
 
@@ -84,6 +84,7 @@ def compact_cache(cache, max_mb):
 
 @require_extension
 def create_segment(channel, series):
+    raise NotImplementedError('no protobuf')
     segment = CacheSegment()
     segment.channelId = channel.id
     segment.index = series.index.astype(np.int64).values.tobytes()
@@ -93,6 +94,7 @@ def create_segment(channel, series):
 
 @require_extension
 def read_segment(channel, bytes):
+    raise NotImplementedError('no protobuf')
     segment = CacheSegment.FromString(bytes)
     index = pd.to_datetime(np.frombuffer(segment.index, np.int64))
     data = np.frombuffer(segment.data, np.double)
