diff --git a/cs/c/configure b/cs/c/configure
index 0f0d4f04c6..d90abd0a26 100755
--- a/cs/c/configure
+++ b/cs/c/configure
@@ -5414,7 +5414,7 @@ do
   esac
 done
 
-SCHEME_CONFIG_ARGS="--machine=${MACH} --disable-x11 ${disable_curses_arg} ${extra_scheme_config_args}"
+SCHEME_CONFIG_ARGS="--machine=${MACH} --disable-x11 --installprefix=/usr --installbin=/usr/bin --installlib=/usr/lib --installman=/usr/share/man --temproot=${D} ${disable_curses_arg} ${extra_scheme_config_args}"
 SCHEME_CROSS_CONFIG_ARGS="--machine=${TARGET_MACH} --disable-x11 ${disable_curses_arg} ${extra_scheme_config_args}"
 
 
diff --git a/cs/c/configure.ac b/racket/src/cs/c/configure.ac
index 9bddef2efc..3813ad8589 100644
--- a/cs/c/configure.ac
+++ b/cs/c/configure.ac
@@ -712,7 +712,7 @@ do
   esac
 done
 
-SCHEME_CONFIG_ARGS="--machine=${MACH} --disable-x11 ${disable_curses_arg} ${extra_scheme_config_args}"
+SCHEME_CONFIG_ARGS="--machine=${MACH} --disable-x11 --installprefix=/usr --installbin=/usr/bin --installlib=/usr/lib --installman=/usr/share/man --temproot=${D} ${disable_curses_arg} ${extra_scheme_config_args}"
 SCHEME_CROSS_CONFIG_ARGS="--machine=${TARGET_MACH} --disable-x11 ${disable_curses_arg} ${extra_scheme_config_args}"
 
 AC_SUBST(CC)
