diff --git a/linkml/__init__.py b/linkml/__init__.py
index 4a9234da..915f564b 100644
--- a/linkml/__init__.py
+++ b/linkml/__init__.py
@@ -1,13 +1,10 @@
 import os
 import sys
 
-import rdflib_shim
 from linkml_runtime.linkml_model import linkml_files
 from linkml_runtime.linkml_model.linkml_files import Format, Source
 from rdflib.plugins.serializers.turtle import TurtleSerializer
 
-shim = rdflib_shim.RDFLIB_SHIM
-
 assert sys.version_info > (
     3,
     7,
