diff --git a/src/Makefile.in b/src/Makefile.in
index 63f909ae14..e161503d0b 100644
--- a/src/Makefile.in
+++ b/src/Makefile.in
@@ -416,7 +416,7 @@ base_obj =
 	bignum.o buffer.o filelock.o insdel.o marker.o \
 	minibuf.o fileio.o dired.o \
 	cmds.o casetab.o casefiddle.o indent.o search.o regex-emacs.o undo.o \
-	alloc.o pdumper.o data.o doc.o editfns.o callint.o \
+	dynlib.o alloc.o pdumper.o data.o doc.o editfns.o callint.o \
 	eval.o floatfns.o fns.o font.o print.o lread.o $(MODULES_OBJ) \
 	syntax.o $(UNEXEC_OBJ) bytecode.o $(COMP_OBJ) \
 	process.o gnutls.o callproc.o \
