diff --git a/setup.py b/setup.py
index 890cf71..198d12f 100755
--- a/setup.py
+++ b/setup.py
@@ -10,8 +10,7 @@ setup(
     packages = ['zdesk'],
     include_package_data = True,
     install_requires = ['requests', 'six'],
-    setup_requires = ['pytest-runner'],
-    tests_require = ['pytest'],
+    tests_require = ['pytest', 'pytest-runner'],
     license='LICENSE.txt',
     url = 'https://github.com/fprimex/zdesk',
     keywords = 'zendesk api helpdesk',
