diff --git a/rdflib/plugin.py b/rdflib/plugin.py
index 2653b188..c1254e36 100644
--- a/rdflib/plugin.py
+++ b/rdflib/plugin.py
@@ -108,6 +108,8 @@ def get(name, kind):
 
 
 try:
+    raise ImportError('pkg_resources is insanely slow see '
+                      'https://github.com/pypa/setuptools/issues/510')
     from pkg_resources import iter_entry_points
 except ImportError:
     pass  # TODO: log a message
diff --git a/rdflib/plugins/sparql/__init__.py b/rdflib/plugins/sparql/__init__.py
index bc1227f2..24ebd1a2 100644
--- a/rdflib/plugins/sparql/__init__.py
+++ b/rdflib/plugins/sparql/__init__.py
@@ -41,6 +41,8 @@ assert operators
 assert parserutils
 
 try:
+    raise ImportError('pkg_resources is insanely slow see '
+                      'https://github.com/pypa/setuptools/issues/510')
     from pkg_resources import iter_entry_points
 except ImportError:
     pass  # TODO: log a message
