diff --git a/funowl/__init__.py b/funowl/__init__.py
index d4c596e..5555932 100644
--- a/funowl/__init__.py
+++ b/funowl/__init__.py
@@ -36,9 +36,6 @@ from . axioms import Axiom
 if sys.version_info < (3, 7, 9):
     warn(f"FunOwl needs python 3.7.9 or later.  Current version: {sys.version_info}")
 
-# Use the rdflib shim
-from rdflib_shim import RDFLIB_SHIM
-x = RDFLIB_SHIM                 # Force import to stay put
 # from rdflib.plugins.serializers.turtle import RecursiveSerializer
 # RecursiveSerializer.roundtrip_prefixes = True
 
diff --git a/requirements.txt b/requirements.txt
index 7ff6aba..c2e4b10 100644
--- a/requirements.txt
+++ b/requirements.txt
@@ -13,6 +13,5 @@
 bcp47
 jsonasobj
 pyjsg>=0.11.6
-rdflib-shim
 rdflib==6.2.0
 rfc3987
